Remove modules and module outputs from state

jbardin opened the pull request:
The OrphanOutputTransformer couldn't find orphaned modules, because it was only looking at the config. Walk the state to find all orphaned outputs and modules.

Create a NodeModuleRemoved and an associated EvalDeleteModule to track
the module in the graph then remove it from the state. The
NodeModuleRemoved dependencies are simply any other node which contains
the module path as a prefix in its path.

This could have probably been done much easier as a step in pruning the
state, but modules are going to have to be promoted to full graph nodes
anyway in order to support count.

apparentlymart commented on terraform/node_module_removed.go:
```diff
@@ -0,0 +1,71 @@
+			}
+			tail := len(state.Modules) - 1
+			state.Modules[i] = state.Modules[tail]
+			state.Modules = state.Modules[:tail]
```
Am I understanding correctly that this moves whichever module is last into the position of the module being deleted?

If so, while indeed the ordering here isn't that important, Terraform Enterprise shows diffs between states and so this reordering may cause some confusion for that diffing engine.

Could we instead retain the relative ordering of all of the other modules by shifting up everything after this point?

```go
copy(state.Modules[i], state.Modules[i+1:])
state.Modules = state.modules[len(state.Modules)-1]
```

I think there may also be a bug here where we skip over one module for each one we remove, since after we've finished closing the gap here (whichever way we do it) there's a different object at `i` which we'll never see, because the next iteration will have already incremented `i`. Perhaps safer and easier to understand here if we were to just construct a new slice altogether and append to it, rather than trying to close this up in-place... this slice should never be large, so copying shouldn't hurt.

jbardin replied:
Yes, that's the most efficient delete pattern.

Good point, even though these may be sorted again later, I'll change this to preserve order just in case. The standard delete preserving order looks like:
```
state.Modules = append(state.Modules[:i], state.Modules[i+1:]...)
```

jbardin replied:
Updated. There no counting bug, since EvalDeleteModule only ever removes a single module.

apparentlymart replied:
Ahh, I see... the `break` in there to make it it stop after one action. Thanks for the clarification!

apparentlymart commented on terraform/state.go:
```diff
@@ -1106,7 +1106,32 @@ func (m *ModuleState) Orphans(c *config.Config) []string {
+
+// OrphanOutputs returns a list of outputs that are in the State but aren't
+// present in the configuration itself.
+func (m *ModuleState) OrphanOutputs(c *config.Config) []string {
```
Could we call this `RemovedOutputs`, as a step towards that terminology change?

jbardin replied:
sure!

apparentlymart commented on terraform/node_module_removed.go:
```diff
@@ -0,0 +1,71 @@
+	defer lock.Unlock()
+
+	// Make sure we have a clean state
+	// Destroyed resources aren't deleted, they're written with an ID of "".
```
:thinking: this seems like a bug in itself! But indeed probably best saved to fix another day...

jbardin replied:
It seems to have been done purposely, and the resource is removed by `prune`. THat's probably best for a later cleanup.

jbardin replied:
This is definitely an issue to research later. 
Pruning the state after every EvalWriteState causes a large number of interpolation errors in tests, where the resource disappears before it can be interpolated. That could totally be from incorrect test fixtures though, but I'm going to postpone cleaning that up until after 0.11.

apparentlymart commented on terraform/context_apply_test.go:
```diff
@@ -2698,10 +2694,7 @@ func TestContext2Apply_moduleOrphanInheritAlias(t *testing.T) {
-module.child:
-  <no state>
-  `)
+	checkStateString(t, state, "")
```
What's the difference between `""` and `"<no state>"` here? (I see `"<no state>"` used in the above example.)

jbardin replied:
It depends on whether there is an empty root module or not, probably depending on when `prune` was called. There are a lot of tests that are relying on very specific behavior for their checks, so I didn't even attempt to go through and audit them all.

apparentlymart replied:
Fair enough! We can iterate on this some more when we revamp the state to be more type-aware, I suppose.

apparentlymart commented on terraform/transform_removed_modules.go:
```diff
@@ -0,0 +1,55 @@
+		}
+	}
+
+	for _, m := range t.State.Modules {
```
I notice that `ModuleState` has a `Dependencies` field whose docstring suggests it should behave similar to that for instances, which seems weird to me since this is the first time we've created a graph node for a module as a whole.

I did some searching in the code and couldn't seem to find anything that actually uses `ModuleState.Dependencies` aside from some generic things like `Equal` and `Copy`, but we seem to have a reasonable amount of test coverage for this even though it doesn't seem to actually ever be used. Do you know what the story is here? Maybe this is vestigial from before the graph builder rewrite in 0.8? :confused: 

Probably we don't actually need to do anything about this, but I just wanted to point it out here in case we should be surfacing those dependencies in the graph somehow. (Probably by implementing `References` on the node rather than directly in here, but still...)

jbardin replied:
I assumed it was to add support for `depends_on`, but didn't look any further into it. 
Yes, we will have to expanded this to work with the ReferenceTransformer in order to be to able to use depends_on or count.

jbardin replied:
Setting up the interfaces to use the ReferenceTransformer may not be as difficult as I initially predicted for this resource. I can try to refactor this early tomorrow, but I don't think we need to hold off the RC if it turns out to be more time consuming than I think.

jbardin commented:
Found the bug in the `ReferenceTransformer` which was preventing the removed module from being ordered properly. 

Removed the references logic from the `RemovedModuleTransformer`, and added the appropriate methods to `NodeModuleRemoved`.
